### DISCUSSION TITLE
Does the Unity version support other formats beside QR-Code?

### DISCUSSION THREAD

**Discussion by dennisvankekem** [micjahn/ZXing.Net#429] [Q&A] [answered]
Hello everyone, 

I'm currently playing around with the unity version of Zxing.net. It works straight out of the box with minimal configuration and it works like a charm.....but only for QR-Codes it seems? 

I've been debugging all night but haven't found any reason as to why. I used a barcode encoder website to test if the problem resides with only code-128 for example, but it seems that only QR-Codes are working, other 2D codes aren't working as well. 

Now, before I'm spending the next week debugging this issue, I'd first like to confirm that it's actually supported. I came across a  similar Github issue on the java version where any other type besides QR-Code was disabled by default in the package lib. 

Just in case, the code is nothing more than:

```
private IBarcodeReader reader = new BarcodeReader();

var result = reader.Decode(cameraImageTexture.GetPixels32(), cameraImageTexture.width, cameraImageTexture.height);

if(result != null){
    // load text result in a TextMeshPro object
    qrCodeText.text = string.Format("result found! {0}", result.Text);
}
```

I've tried to play around with the `reader.Options.PossibleFormats`, `reader.Options.TryHarder` etc but no luck.  

Can someone confirm if the question in the title is true, or is there something obvious that I'm missing? 

Edit 1: 

Example of barcode that I expect to be scannable: 

Code-128:
<img width="375" alt="image" src="https://user-images.githubusercontent.com/38350840/175276308-f096ad73-6bad-48f9-a4ce-dea8166becee.png">

Code-39:
<img width="400" alt="image" src="https://user-images.githubusercontent.com/38350840/175276638-f36f8b9c-5f80-4df2-8e93-95f9e0d7a416.png">

[source](https://barcode.tec-it.com/en/Code93?data=ABC-1234-%2F%2B)

Edit 2: 

I still haven't figured out what causes this problem. I do downsize the image by 50% as recommended. See:

```
var conversionParams = new XRCpuImage.ConversionParams {
    // Get the entire image.
    inputRect = new RectInt(0, 0, image.width, image.height),

    // Downsample by 2.
    outputDimensions = new Vector2Int(image.width / 2, image.height / 2),

    // Choose RGBA format.
    outputFormat = TextureFormat.RGBA32,

    // Flip across the vertical axis (mirror image).
    transformation = XRCpuImage.Transformation.MirrorY
};
```

**Comment by micjahn** [comment]
ZXing.Net supports more than only QR codes.
If you still have an issue here, please provide some sample images which doesn't work and the content which is encoded in the barcode (if possible).

**Reply by dennisvankekem** [reply]
Thanks for taking some time to look into my issue, I appreciate it a lot. I've updated my question with two example barcodes that fail to work on my device(Unity Android). The barcodes being code-128 and code-39. I've used the TEC-IT barcode generator(source is below the examples). The text below the barcodes is the result text that I expected to appear after scanning said barcodes.

I've also updated my question with the conversion parameters that I use to convert the camera feed to monochrome images for ZXing. These conversion parameters however are the default XRCpuImage conversion parameters that Unity recommends, see: [this link to docs](https://docs.unity3d.com/Packages/com.unity.xr.arfoundation@5.0/manual/cpu-camera-image.html#synchronously-convert-to-grayscale-and-color).

If you need anything else, please let me know. I will take the weekend off to make a sandbox with just ZXing and the ARfoundation setup to see if I can narrow down the issue.

**Reply by dennisvankekem** [reply]
Version used is: 0.16.8

**Reply by dennisvankekem** [reply]
Okay, so the problem has to do with the default rotation. I added the converted textures to a debug feed and it seems that by default the texture is rotated by 90 degrees to the right. I have no idea as to why this is as everything in my project has a rotation value of 0,0,0 (quaternion.identity).

If I disable landscape mode and rotate my phone by 90 degrees it does successfully scan the barcodes. 

I will find out what causes the problem and report back if I find anything useful.

**Comment by dennisvankekem** [comment, marked as answer]
My sincere apologies, the issue is not with ZXing but rather with the way that the AR foundation handles raw camera output. Apparently, when you access the camera on a mobile device the default camera rotation is landscape. So even if your phone is in portrait mode, the camera's raw image output is still in landscape. The AR foundation does not care for device rotation, however when your camera is in portrait mode it still shows it as portrait, while in the background your camera images are processed in landscape. This is a huge false positive understandably confusing for a lot of developers. To fix the issue you can rotate your Texture2D with this code:

```
Texture2D rotateTexture(Texture2D originalTexture, bool clockwise)
{
   Color32[] original = originalTexture.GetPixels32();
   Color32[] rotated = new Color32[original.Length];
   int w = originalTexture.width;
   int h = originalTexture.height;

   int iRotated, iOriginal;

   for (int j = 0; j < h; ++j)
   {
       for (int i = 0; i < w; ++i)
       {
           iRotated = (i + 1) * h - j - 1;
           iOriginal = clockwise ? original.Length - 1 - (j * w + i) : j * w + i;
           rotated[iRotated] = original[iOriginal];
       }
   }

   Texture2D rotatedTexture = new Texture2D(h, w);
   rotatedTexture.SetPixels32(rotated);
   rotatedTexture.Apply();
   return rotatedTexture;
}
```

@micjahn if people in the future come across the same problem you might want to redirect them to this topic.

**Reply by micjahn** [reply]
Thanks for the clarification
